"""
Rules for indexing, sorting, and modifying BAM files

Uses:
  - sambamba
  - samtools

Parameters:
  - bam_sort_memory
"""

__author__  = "Sean Davis <http://watson.nci.nih.gov/~sdavis>"
__license__ = "MIT"

rule index_bam:
    threads: 8
    input: 
        "{base}.bam"
    output: 
        "{base}.bam.bai"
    params: batch="-l nodes=1:gpfs"
    shell: "module load sambamba && sambamba index -t {threads} {input}"

rule indelrealigner:
    input: bam="{base}.bam",
           intervals="{base}.intervals",
           bai='{base}.bam.bai'
    output: bam=temp("{base}.realigned.bam")
    params: batch="-l nodes=1:gpfs"
    threads: 1
    shell: "do this"
 
#shell(gatk.indelrealigner(input.bam,output.bam,input.intervals,indelvcf='/data/CCRBioinfo/public/GATK/bundle/2.3/hg19/Mills_and_1000G_gold_standard.indels.hg19.vcf'))

rule realignertargetcreator:
    input: bam="{base}.bam",
           bai="{base}.bam.bai"
    output: "{base}.intervals"
    params: batch="-l nodes=1:gpfs"
    threads: 32
    shell: "do this"

#shell(gatk.realigntargetcreator(input[0],output,indelvcf='/data/CCRBioinfo/public/GATK/bundle/2.3/hg19/Mills_and_1000G_gold_standard.indels.hg19.vcf',mem='16g',threads=threads))





rule markdups:
    threads: 1
    input: '{base}.bam','{base}.bam.bai'
    output: bam='{base}.md.bam',metrics='{base}.dupmetrics'
    params: batch="-l nodes=1:gpfs"
    version: "1.119"
    shell: "module load picard/{version} && "
        "java -jar $PICARDJARPATH/MarkDuplicates.jar "
        "AS=true M={output.metrics} O={output.bam} "
        "I={input[0]} REMOVE_DUPLICATES=false VALIDATION_STRINGENCY=SILENT"
